### PR TITLE
NEW: Management of extrafields on customer prices, level prices and default prices

### PR DESCRIPTION
**Pull request by altairis-noe**
# NEW: Management of extrafields on customer prices, level prices and default prices

## Add configuration of extrafields on customer prices and price level for products

image

## Add extrafields on customer prices in the product Sales price tab and for the default price of the product

image

## Add of extrafields on customer prices on the thirdparty Customer prices tab

image

Extrafields are managed while creating and updating customer prices, price level and default prices.

### CODE REVIEW

File: htdocs/core/class/extrafields.class.php [hunk 1]
```diff
@@ -956,7 +956,8 @@ private function update_label($attrname, $label, $type, $size, $elementtype, $un
...
 	 * 	@param	boolean		$forceload			Force load of extra fields whatever is status of cache.
 	 *  @param  string		$attrname           The name of the attribute.
-	 *  @return array{}|array{label:array<string,string>,type:array<string,string>,size:array<string,string>,default:array<string,string>,computed:array<string,string>,unique:array<string,int>,required:array<string,int>,param:array<string,mixed>,perms:array<string,mixed[]>,list:array<string,int>|array<string,string>,pos:array<string,int>,totalizable:array<string,int>,help:array<string,string>,printable:array<string,int>,enabled:array<string,int>,langfile:array<string,string>,css:array<string,string>,csslist:array<string,string>,hidden:array<string,int>,mandatoryfieldsofotherentities?:array<string,string>,loaded?:int,count:int}		Array of attributes keys+label for all extra fields.  Note: count set as present to avoid static analysis notices
+	 *  @return array<string>					Array of attributes keys+label for all extra fields.  Note: count set as present to avoid static analysis notices
```
Comment (mdeweerd) [inline htdocs/core/class/extrafields.class.php:959]: IMHO the new return value is array<string,string> .
If not specifed as such, there will be notices indicating that an array key to access the result is a string while an int is expected.

Comment (altairis-noe) [reply htdocs/core/class/extrafields.class.php]: good point, I corrected!

File: htdocs/install/mysql/tables/llx_product_customer_price_extrafields.key.sql [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU General Public License for more details.
+--
+-- You should have received a copy of the GNU General Public License
+-- along with this program. If not, see <http://www.gnu.org/licenses/>.
+--
+-- ===================================================================
+
+
+ALTER TABLE llx_product_customer_price_extrafields ADD UNIQUE INDEX uk_product_customer_price_extrafields (fk_object);
```
Comment (eldy) [inline htdocs/install/mysql/tables/llx_product_customer_price_extrafields.key.sql:21]: To follow contributions rules, thanks to push first the database addition into a separated PR ?

**Comment by altairis-noe** [comment]
Right @eldy #31395 is live!

Should we start this one anew when #31395 is accepted?

**Comment by altairis-noe** [comment]
Hey there @eldy. I think we are good to go, here 🙂

**Pull request merged**